File: src/presentation/tab_view.py
from java.awt import BorderLayout
from java.awt import Dimension
from java.awt import FlowLayout
from javax.swing import JButton
from javax.swing import JLabel
from javax.swing import JPanel
from javax.swing import JScrollPane
from javax.swing import JTable
from javax.swing import JTextField

from presentation.table_model import EndpointTableModel


class TabView(object):
    def __init__(self):
        self.root_panel = JPanel(BorderLayout())

        self.scan_button = JButton("Scan")
        self.stop_button = JButton("Stop")
        self.search_label = JLabel("Search:")
        self.search_field = JTextField(24)
        self.export_button = JButton("Export CSV")
        self.status_label = JLabel("Ready")

        self.table_model = EndpointTableModel()
        self.result_table = JTable(self.table_model)
        self.result_table.setFillsViewportHeight(True)
        self.result_table.setAutoCreateRowSorter(True)

        self._build_layout()

    def _build_layout(self):
        top_panel = JPanel(FlowLayout(FlowLayout.LEFT))
        top_panel.add(self.scan_button)
        top_panel.add(self.stop_button)
        top_panel.add(self.search_label)
        top_panel.add(self.search_field)
        top_panel.add(self.export_button)

        table_scroll = JScrollPane(self.result_table)
        table_scroll.setPreferredSize(Dimension(900, 420))

        bottom_panel = JPanel(BorderLayout())
        bottom_panel.add(self.status_label, BorderLayout.WEST)

        self.root_panel.add(top_panel, BorderLayout.NORTH)
        self.root_panel.add(table_scroll, BorderLayout.CENTER)
        self.root_panel.add(bottom_panel, BorderLayout.SOUTH)

    def get_root_component(self):
        return self.root_panel

    def set_status(self, message):
        text = message if message else ""
        self.status_label.setText(text)

    def set_records(self, records):
        self.table_model.set_records(records)
        self.set_status("Records: %d" % self.table_model.getRowCount())

    def clear_records(self):
        self.table_model.clear()
        self.set_status("Records: 0")

    def get_search_keyword(self):
        return self.search_field.getText().strip()

    def get_components(self):
        return {
            "scan_button": self.scan_button,
            "stop_button": self.stop_button,
            "search_field": self.search_field,
            "export_button": self.export_button,
            "status_label": self.status_label,
            "result_table": self.result_table,
            "table_model": self.table_model,
        }

File: src/main/java/com/secuworm/endpointcollector/presentation/TabView.java
package com.secuworm.endpointcollector.presentation;

import com.secuworm.endpointcollector.domain.EndpointRecord;

import javax.swing.AbstractAction;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTable;
import javax.swing.JTextField;
import javax.swing.KeyStroke;
import javax.swing.ListSelectionModel;
import java.awt.event.ActionEvent;
import java.awt.event.InputEvent;
import java.awt.event.KeyEvent;
import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.util.ArrayList;
import java.util.List;

public class TabView {
    private static final String LICENSE_TEXT = "Copyright © 2026 SECUWORM. All rights reserved.";

    private final JPanel rootPanel;
    private final JTextField searchField;
    private final JCheckBox regexCheckBox;
    private final JButton exportButton;
    private final JButton sendToRepeaterButton;
    private final JLabel statusLabel;
    private final EndpointTableModel tableModel;
    private final JTable resultTable;

    public TabView() {
        rootPanel = new JPanel(new BorderLayout());
        searchField = new JTextField(24);
        regexCheckBox = new JCheckBox("Regex");
        exportButton = new JButton("Export CSV");
        sendToRepeaterButton = new JButton("Send to Repeater");
        statusLabel = new JLabel("Ready");
        tableModel = new EndpointTableModel();
        resultTable = new JTable(tableModel);
        resultTable.setFillsViewportHeight(true);
        resultTable.setAutoCreateRowSorter(true);
        resultTable.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);

        JPanel topPanel = new JPanel(new BorderLayout());

        JPanel leftPanel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        leftPanel.add(new JLabel("Search:"));
        leftPanel.add(searchField);
        leftPanel.add(regexCheckBox);
        leftPanel.add(exportButton);
        leftPanel.add(sendToRepeaterButton);

        JPanel rightPanel = new JPanel(new FlowLayout(FlowLayout.RIGHT));
        rightPanel.add(new JLabel(LICENSE_TEXT));

        topPanel.add(leftPanel, BorderLayout.WEST);
        topPanel.add(rightPanel, BorderLayout.EAST);

        JScrollPane tableScroll = new JScrollPane(resultTable);
        tableScroll.setPreferredSize(new Dimension(900, 420));

        JPanel bottomPanel = new JPanel(new BorderLayout());
        bottomPanel.add(statusLabel, BorderLayout.WEST);

        rootPanel.add(topPanel, BorderLayout.NORTH);
        rootPanel.add(tableScroll, BorderLayout.CENTER);
        rootPanel.add(bottomPanel, BorderLayout.SOUTH);

        bindRepeaterShortcut();
    }

    public JPanel getRootComponent() {
        return rootPanel;
    }

    public JTextField getSearchField() {
        return searchField;
    }

    public JButton getExportButton() {
        return exportButton;
    }

    public JCheckBox getRegexCheckBox() {
        return regexCheckBox;
    }

    public JButton getSendToRepeaterButton() {
        return sendToRepeaterButton;
    }

    public void setStatus(String message) {
        statusLabel.setText(message == null ? "" : message);
    }

    public void setRecords(List<EndpointRecord> records) {
        tableModel.setRecords(records);
        setStatus("Records: " + tableModel.getRowCount());
    }

    public String getSearchKeyword() {
        return searchField.getText() == null ? "" : searchField.getText().trim();
    }

    public boolean isRegexSearchEnabled() {
        return regexCheckBox.isSelected();
    }

    public EndpointRecord getSelectedRecord() {
        int selectedViewRow = resultTable.getSelectedRow();
        if (selectedViewRow < 0) {
            return null;
        }
        int selectedModelRow = resultTable.convertRowIndexToModel(selectedViewRow);
        return tableModel.getRecord(selectedModelRow);
    }

    public List<EndpointRecord> getSelectedRecords() {
        int[] selectedViewRows = resultTable.getSelectedRows();
        List<EndpointRecord> selectedRecords = new ArrayList<>();
        if (selectedViewRows == null || selectedViewRows.length == 0) {
            return selectedRecords;
        }
        for (int selectedViewRow : selectedViewRows) {
            int selectedModelRow = resultTable.convertRowIndexToModel(selectedViewRow);
            EndpointRecord record = tableModel.getRecord(selectedModelRow);
            if (record != null) {
                selectedRecords.add(record);
            }
        }
        return selectedRecords;
    }

    private void bindRepeaterShortcut() {
        KeyStroke shortcut = KeyStroke.getKeyStroke(KeyEvent.VK_R, InputEvent.CTRL_DOWN_MASK);
        String actionKey = "send-to-repeater";

        AbstractAction action = new AbstractAction() {
            @Override
            public void actionPerformed(ActionEvent event) {
                sendToRepeaterButton.doClick();
            }
        };

        rootPanel.getInputMap(JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT).put(shortcut, actionKey);
        rootPanel.getActionMap().put(actionKey, action);
    }
}
